django/middleware/csrf.py

```diff
@@ -437,15 +437,16 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
         return self._accept(request)
 
     def process_response(self, request, response):
-        if not getattr(request, 'csrf_cookie_needs_reset', False):
-            if getattr(response, 'csrf_cookie_set', False):
-                return response
-
-        if not request.META.get("CSRF_COOKIE_USED", False):
+        # Prevent _set_token() from being called twice (e.g. when both a
+        # decorator and middleware are used).
+        if getattr(response, 'csrf_cookie_set', False):
```

Comment:
I cannot think of any legitimate reason on how you could trigger this sensibly to go wrong, but do imagine the following (order):

 * csrf middleware
 * custom middleware
 * decorator 
 
 Assuming "custom middleware" calls `rotate_token` (in `process_response` or so) which sets `csrf_cookie_needs_reset` to True but the decorator already did set `csrf_cookie_set` to True as well, then the token wouldn't get updated.
 
 The code (not nice, but I couldn't come up with anything better to express what I ment) would look like this:
 ```py
     def process_response(self, request, response):
        # The cookie needs to be reset, reset it but also reset the flag so we do it not more often than neccessary (can still be more than once per request)
        if getattr(request, 'csrf_cookie_needs_reset', False):
            self.csrf_cookie_needs_reset = False
        elif getattr(response, 'csrf_cookie_set', False):
            return response
        elif not request.META.get("CSRF_COOKIE_USED", False):
            return response

        # Set the CSRF cookie even if it's already set, so we renew
        # the expiry timer.
        self._set_token(request, response)
        response.csrf_cookie_set = True
        return response
```

Not trying to bike-shed here, I honestly do not know how `process_response` is supposed to behave in such edge-cases (or whether they even make sense at all)

Reply:
That's a great thought, @apollo13. Thanks!

I will note that prior to PR #5605 (so prior to Django 1.10), I don't think the scenario you describe would have been supported. See below for what `process_response()` looked like at that time (where `csrf_processing_done` is playing the role of `csrf_cookie_set`). The logic back then is basically what I'm proposing now:
https://github.com/django/django/blob/6d9c5d46e644a8ef93b0227fc710e09394a03992/django/middleware/csrf.py#L221-L241

However, I don't see any reason why we can't support that new scenario. Let me add a test case for that situation, and we can see how it looks.